Add dockermake.bsh for a generic docker image build system for drone.

wrouesnel opened the pull request:
Re-PR of #23 which I accidentally merged when pushing 22.

This PR adds the dockermake build script and tests for correct operation.

stephanh commented on dockermake.bsh:
```diff
@@ -0,0 +1,109 @@
+}
+
+# Options!
+case $ACTION in
```
Duplicate of line 25+.

stephanh replied:
Ok. Not really a duplicate but still a bit confusing that we have this switch in two places. Can we combine them?

stephanh commented on dockermake.bsh:
```diff
@@ -0,0 +1,109 @@
+# Make might be semantically pretty, but the language sucks. This is much
+# clearer.
+
+# Build arguments to pass to docker build
```
Can you add a bit more of comment to this file. In particular info such as:

* expects VERSION file
* Can provide TAG as env `TAG`.

Etc.

stephanh commented on dockermake.bsh:
```diff
@@ -0,0 +1,109 @@
+    done
+}
+
+do_push() {
```
Can we add branch detection for this. We only want to do a push if we are on a release branch (master or other) and not on a pr.

stephanh replied:
This is probably a function that could go into the ci lib.

wrouesnel replied:
We always have to push because there's no guarantee that additional tests will be scheduled on the same node. Version numbers are unique under the scheme.

stephanh replied:
Can you expand on that. What additional tests are you referring to? I don't think we want to publish a new docker image for every time we update a PR.

wrouesnel replied:
If you build image A, which B and C derive from, then on pull request we actually want to build B and C as well to find out whether we're break a whole chain of builds. B and C can take place on completely unrelated nodes, which means the pull request image needs to go into the shared store.

In a practical sense, this is all about how you structure the build file - i.e. you can set a deploy hook which only runs when the branch is master. The script really shouldn't be making this decision.

stephanh replied:
Ok. Unfortunately atm we don't really have that kind of build chaining and it is something we would have to add on top of drone/travis (TeamCity has support for it). 

In the mean time to be consistent with the existing scripts and to not send lots and lots of docker images to artifactory can we add the check to the script?

When our CI pipeline matures we need to revisit this.

stephanh replied:
Have a look at `sbt-deploy-to`.

stephanh commented on dockermake.bsh:
```diff
@@ -0,0 +1,109 @@
+# clearer.
+
+# Build arguments to pass to docker build
+BUILD_ARGS=()
```
set -e ?

wrouesnel replied:
It's mixed whether `set -e` is good practice or not. It's bash, so test coverage is more reliable - and also I've made some changes where we don't want that, because we want to see how much of the build fails then exit failure.

wrouesnel commented:
Ok so I've had to refactor to support build arguments in a generic way (otherwise we end up needing to reimplement the `docker` command line parser). The tests have been updated and simplified to be more maintainable (and produce less unnecessarily verbose output).

`RELEASE_BRANCHES` is now defaulted to `master`, and checks `CI_BRANCH`.

I've also introduced `Array_Contains` to cut down on code duplication in `lib-ci`.

stephanh commented:
:shipit:
